Libraries/LibC/qsort.cpp

```diff
@@ -24,103 +24,65 @@
- * Redistribution and use in source and binary forms, with or without
- * modification, are permitted provided that the following conditions
- * are met:
- * 1. Redistributions of source code must retain the above copyright
- *    notice, this list of conditions and the following disclaimer.
- * 2. Redistributions in binary form must reproduce the above copyright
- *    notice, this list of conditions and the following disclaimer in the
- *    documentation and/or other materials provided with the distribution.
- * 3. All advertising materials mentioning features or use of this software
- *    must display the following acknowledgement:
- *	This product includes software developed by the University of
- *	California, Berkeley and its contributors.
- * 4. Neither the name of the University nor the names of its contributors
- *    may be used to endorse or promote products derived from this software
- *    without specific prior written permission.
- *
- * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
- * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
- * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
- * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
- * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
- * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
- * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
- * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
- * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
- * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
- * SUCH DAMAGE.
- */
-
 #if defined(LIBC_SCCS) && !defined(lint)
-static char sccsid[] = "@(#)qsort.c	5.9 (Berkeley) 2/23/91";
+static char sccsid[] = "@(#)qsort.cpp	1.0 (Serenity) 2020-06-30";
 #endif /* LIBC_SCCS and not lint */
 
+#include <AK/QuickSort.h>
 #include <stdlib.h>
 #include <sys/types.h>
 
-static void insertion_sort(void* bot, size_t nmemb, size_t size, int (*compar)(const void*, const void*));
-static void insertion_sort_r(void* bot, size_t nmemb, size_t size, int (*compar)(const void*, const void*, void*), void* arg);
+class SizedObjectIterator {
+public:
+    SizedObjectIterator() = delete;
+    SizedObjectIterator(void *pos, size_t element_size) : m_pos(static_cast<char*>(pos)), m_element_size(element_size) {}
+    bool operator!=(const SizedObjectIterator& other) { return m_pos != other.m_pos; }
+    bool operator==(const SizedObjectIterator& other) { return m_pos == other.m_pos; }
+    bool operator<(const SizedObjectIterator& other) { return m_pos < other.m_pos; }
+    bool operator>(const SizedObjectIterator& other) { return m_pos > other.m_pos; }
+    bool operator<=(const SizedObjectIterator& other) { return m_pos <= other.m_pos; }
+    bool operator>=(const SizedObjectIterator& other) { return m_pos >= other.m_pos; }
```

Comment:
Out of curiousity, can you use the <=> operator here? I think gcc 10 supports it.